packages/react-native-renderer/src/ReactNativeBridgeEventPlugin.js

```diff
@@ -38,13 +47,21 @@ function accumulateDirectionalDispatches(inst, phase, event) {
       console.error('Dispatching inst must not be null');
     }
   }
-  const listener = listenerAtPhase(inst, event, phase);
-  if (listener) {
+  const listeners = listenersAtPhase(
+    inst,
+    event,
+    phase,
+    event instanceof CustomEvent,
+  );
+  if (listeners && listeners.length > 0) {
     event._dispatchListeners = accumulateInto(
       event._dispatchListeners,
-      listener,
+      listeners,
```

Comment:
The vast majority case is going to be one listener. It feels a bit wrong to me that we're now paying the runtime cost assuming there's more than one. I.e. always having an array.

Maybe we can simplify the code. E.g. `accumulateInto` here is a helper that tries to add `T | Array<T>` on the right side to `T | Array<T>` on the left side. But now our right side is always an array. So the `isArray` check inside of `accumulateInto` is pointless. And creating an intermediate array only to stuff it back into another array also seems pointless.

I would suggest that if we're doing this, let's get rid of `accumulateInto` altogether and just write it as a for loop. Instead of `listenersAtPhase` that always allocates, you could have something like `pushListenersAtPhase` which will lazily create `event._dispatchListeners` if needed, and then `push` into it. No `accumulateInto`, no intermediate checks, no extra arrays. It already has the access to `event`, anyway.

Reply:
Tried to address this in the next PR by returning T | Array<T> | null from getEventListeners and avoid allocations more aggressively